src/internal/log/http.go

```diff
@@ -44,7 +46,43 @@ func AddLoggerToHTTPServer(rctx context.Context, name string, s *http.Server) {
 			ctx = ChildLogger(ctx, "", WithFields(id))
 			ctx = metadata.NewOutgoingContext(ctx, metadata.MD{"x-request-id": requestID})
 			r = r.WithContext(ctx)
+
+			var respStatusCode int
+			var errbuf []byte
+			w = httpsnoop.Wrap(w, httpsnoop.Hooks{
+				WriteHeader: func(next httpsnoop.WriteHeaderFunc) httpsnoop.WriteHeaderFunc {
+					return func(code int) {
+						respStatusCode = code
+						next(code)
+					}
+				},
+				Write: func(next httpsnoop.WriteFunc) httpsnoop.WriteFunc {
+					return func(b []byte) (int, error) {
+						if respStatusCode == 0 {
+							respStatusCode = 200
+						}
+						if respStatusCode >= 400 {
+							if errbuf == nil {
+								// TODO would it be better to statically alloc this? (e.g.
+								// [1000]byte above?)
+								errbuf = make([]byte, 0, 1000)
+							}
+							if len(errbuf) < cap(errbuf) {
+								m := len(errbuf)
+								n := min(len(b), cap(errbuf)-m)
+								errbuf = errbuf[:m+n] // extend errbuf
+								copy(errbuf[m:], b[:n])
+							}
+						}
+						return next(b)
+					}
+				},
+			})
+
 			orig.ServeHTTP(w, r)
+			if ua := r.Header.Get("user-agent"); ua != "Envoy/HC" {
```

Comment:
Let's compute this conditional once and do the `if !healthCheck` in the two places that care.